<a href="https://colab.research.google.com/github/Somrat390/Employee-Attrition-Prediction-Platform/blob/main/Notebooks/1_business_understanding_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem Statement

## Business Problem

Employee attrition refers to the voluntary or involuntary departure of employees from an organization. High attrition rates can negatively impact a company's productivity, increase recruitment and training costs, reduce team morale, and result in the loss of valuable knowledge and experience.

Replacing an employee is both time-consuming and expensive. Organizations invest significant resources in hiring, onboarding, and training employees. When experienced employees leave unexpectedly, these investments are lost, and business operations may be disrupted.

The objective of this project is to develop a machine learning model that predicts whether an employee is likely to leave the company based on demographic, professional, and workplace-related factors. By identifying employees who are at a high risk of attrition, Human Resources (HR) teams can take proactive retention measures such as career development opportunities, salary adjustments, workload balancing, or employee engagement programs.

This project aims to transform historical employee data into actionable insights that support data-driven decision-making and improve employee retention strategies.

# Target Variable

## The target variable in this dataset is Attrition.

Yes (1): The employee has left the company.
No (0): The employee is still working at the company.

Since the goal is to predict one of two possible outcomes (Leave or Stay), this is a Binary Classification problem.



# Project Objective

The primary objective of this project is to build an end-to-end Employee Attrition Prediction Platform capable of estimating the probability that an employee will leave the organization. In addition to generating predictions, the platform will provide interpretable insights into the key factors influencing attrition, enabling HR professionals to make informed and proactive retention decisions.

# Success Metrics

Since employee attrition datasets are typically imbalanced, relying solely on Accuracy can be misleading. For example, if most employees stay with the company, a model may achieve high accuracy simply by predicting that every employee will stay, while failing to identify those who are actually at risk of leaving.

Therefore, this project will focus on evaluation metrics that provide a more reliable assessment of model performance.

## Recall (Primary Metric)

Recall measures the proportion of employees who actually left the company that were correctly identified by the model.

A high Recall is important because failing to identify an employee who is likely to leave (False Negative) may result in unexpected resignations, increased recruitment costs, and loss of experienced talent. Therefore, maximizing Recall helps HR intervene before valuable employees resign.

## F1 Score

The F1 Score is the harmonic mean of Precision and Recall. It provides a balanced evaluation when both False Positives and False Negatives are important.

In this project, an excessively high number of False Positives would cause HR to spend unnecessary time and resources on employees who are not actually at risk, while too many False Negatives would allow high-risk employees to leave unnoticed. The F1 Score helps maintain a balance between these two types of errors.

## ROC-AUC Score

The Receiver Operating Characteristic - Area Under the Curve (ROC-AUC) measures the model's ability to distinguish between employees who are likely to leave and those who are likely to stay across different decision thresholds.

A higher ROC-AUC score indicates that the model has a stronger ability to rank high-risk employees above low-risk employees, making it a robust metric for evaluating classification performance independent of a single probability threshold.

# Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load The Dataset

In [2]:
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

# Basic Exploration of the Dataset

In [4]:
df.shape

(1470, 35)

In [5]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [6]:
df.tail()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1465,36,No,Travel_Frequently,884,Research & Development,23,2,Medical,1,2061,...,3,80,1,17,3,3,5,2,0,3
1466,39,No,Travel_Rarely,613,Research & Development,6,1,Medical,1,2062,...,1,80,1,9,5,3,7,7,1,7
1467,27,No,Travel_Rarely,155,Research & Development,4,3,Life Sciences,1,2064,...,2,80,1,6,0,3,6,2,0,3
1468,49,No,Travel_Frequently,1023,Sales,2,3,Medical,1,2065,...,4,80,0,17,3,2,9,6,0,8
1469,34,No,Travel_Rarely,628,Research & Development,8,3,Medical,1,2068,...,1,80,0,6,3,4,4,3,1,2


In [18]:
df.sample(5, random_state=42)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1041,28,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,...,4,80,0,6,4,3,5,4,1,3
184,53,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,...,3,80,2,5,3,3,4,2,1,3
1222,24,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,...,3,80,1,1,2,3,1,0,0,0
67,45,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,...,3,80,1,25,2,3,1,0,0,0
220,36,No,Travel_Rarely,1396,Research & Development,5,2,Life Sciences,1,304,...,4,80,0,16,3,4,13,11,3,7


In [17]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [9]:
df.describe()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,36.923810,802.485714,9.192517,2.912925,1.0,1024.865306,2.721769,65.891156,2.729932,2.063946,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,9.135373,403.509100,8.106864,1.024165,0.0,602.024335,1.093082,20.329428,0.711561,1.106940,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,491.250000,2.000000,48.000000,2.000000,1.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,802.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


In [10]:
df.describe(include='object')

,Attrition,BusinessTravel,Department,EducationField,Gender,JobRole,MaritalStatus,Over18,OverTime
count,1470,1470,1470,1470,1470,1470,1470,1470,1470
unique,2,3,3,6,2,9,3,1,2
top,No,Travel_Rarely,Research & Development,Life Sciences,Male,Sales Executive,Married,Y,No
freq,1233,1043,961,606,882,326,673,1470,1054


In [15]:
df.memory_usage(deep=True).sum() /(1024 ** 2)

np.float64(1.0161409378051758)

## Dataset Summary

- **Number of Rows:** 1,470
- **Number of Columns:** 35
- **Categorical Features:** 9
- **Numerical Features:** 26
- **Dataset Memory Usage:** Approximately 1 MB

### Initial Observations

- The dataset contains both numerical and categorical features.
- The target variable is **Attrition**, indicating whether an employee has left the company.
- The dataset appears to be well-structured and suitable for supervised binary classification.
- No conclusions regarding missing values or duplicates are made at this stage; these will be investigated during the data quality assessment.

| Feature                  | Type                        | Meaning                                                                                 |
| ------------------------ | --------------------------- | --------------------------------------------------------------------------------------- |
| Age                      | Numeric                     | Age of the employee (in years).                                                         |
| Attrition                | Target (Binary Categorical) | Whether the employee left the company (Yes/No).                                         |
| BusinessTravel           | Nominal                     | Frequency of business travel (Non-Travel, Travel Rarely, Travel Frequently).            |
| DailyRate                | Numeric                     | Daily salary rate of the employee.                                                      |
| Department               | Nominal                     | Department where the employee works.                                                    |
| DistanceFromHome         | Numeric                     | Distance between the employee's home and workplace (miles).                             |
| Education                | Ordinal                     | Education level (1 = Below College, 2 = College, 3 = Bachelor, 4 = Master, 5 = Doctor). |
| EducationField           | Nominal                     | Field of study of the employee.                                                         |
| EmployeeCount            | Numeric                     | Number of employees (constant value).                                                   |
| EmployeeNumber           | Identifier                  | Unique employee identifier.                                                             |
| EnvironmentSatisfaction  | Ordinal                     | Employee satisfaction with the work environment (1 = Low, 4 = Very High).               |
| Gender                   | Nominal                     | Gender of the employee.                                                                 |
| HourlyRate               | Numeric                     | Hourly salary rate.                                                                     |
| JobInvolvement           | Ordinal                     | Employee involvement in their job (1 = Low, 4 = Very High).                             |
| JobLevel                 | Ordinal                     | Employee job level within the company.                                                  |
| JobRole                  | Nominal                     | Employee's job role.                                                                    |
| JobSatisfaction          | Ordinal                     | Overall job satisfaction (1 = Low, 4 = Very High).                                      |
| MaritalStatus            | Nominal                     | Marital status of the employee.                                                         |
| MonthlyIncome            | Numeric                     | Monthly income of the employee.                                                         |
| MonthlyRate              | Numeric                     | Monthly salary rate.                                                                    |
| NumCompaniesWorked       | Numeric                     | Number of companies the employee has worked for previously.                             |
| Over18                   | Nominal                     | Indicates whether the employee is over 18 years old (constant value).                   |
| OverTime                 | Nominal                     | Whether the employee works overtime (Yes/No).                                           |
| PercentSalaryHike        | Numeric                     | Percentage increase in salary from the previous year.                                   |
| PerformanceRating        | Ordinal                     | Performance evaluation rating (1 = Low, 4 = Outstanding).                               |
| RelationshipSatisfaction | Ordinal                     | Satisfaction with workplace relationships (1 = Low, 4 = Very High).                     |
| StandardHours            | Numeric                     | Standard working hours (constant value).                                                |
| StockOptionLevel         | Ordinal                     | Employee stock option level.                                                            |
| TotalWorkingYears        | Numeric                     | Total years of professional work experience.                                            |
| TrainingTimesLastYear    | Numeric                     | Number of training sessions attended in the previous year.                              |
| WorkLifeBalance          | Ordinal                     | Employee work-life balance (1 = Bad, 4 = Best).                                         |
| YearsAtCompany           | Numeric                     | Total years worked at the current company.                                              |
| YearsInCurrentRole       | Numeric                     | Years spent in the current job role.                                                    |
| YearsSinceLastPromotion  | Numeric                     | Years since the employee's last promotion.                                              |
| YearsWithCurrManager     | Numeric                     | Years working with the current manager.                                                 |


# Feature Categorization

## Identifier Columns
* EmployeeNumber (Unique identifier for each employee; not useful for prediction.)

## Constant Columns

These columns contain only one unique value and do not provide any predictive information.

* EmployeeCount
* Over18
* StandardHours

These will likely be removed during preprocessing.

# High Cardinality Columns

This dataset does not contain any truly high-cardinality categorical features.

#### The feature with the highest number of categories is:

#### JobRole (9 categories)

Since it has only a small number of unique values, it is not considered high cardinality.

# Ordinal Features

These features have a natural order.

* Education
* EnvironmentSatisfaction
* JobInvolvement
* JobLevel
* JobSatisfaction
* PerformanceRating
* RelationshipSatisfaction
* StockOptionLevel
* WorkLifeBalance

# Nominal Features

These features have no inherent order.

* Attrition
* BusinessTravel
* Department
* EducationField
* Gender
* JobRole
* MaritalStatus
* Over18
* OverTime

## Initial Feature Assessment

- EmployeeNumber is a unique identifier and is not expected to contribute to predicting employee attrition.
- EmployeeCount, Over18, and StandardHours appear to have constant values and are likely to be removed during preprocessing.
- Several features, such as Education, JobLevel, JobSatisfaction, and WorkLifeBalance, are ordinal and should be encoded while preserving their natural order.
- Most categorical variables are nominal and will likely require One-Hot Encoding.
- No high-cardinality categorical features are present, simplifying the encoding process.

# Intial Data Quality Check

In [19]:
df.isnull().sum()

,0
Age,0
Attrition,0
BusinessTravel,0
DailyRate,0
Department,0
DistanceFromHome,0
Education,0
EducationField,0
EmployeeCount,0
EmployeeNumber,0


## Missing Values

The dataset contains **no missing values** across all 35 features. Therefore, no imputation techniques will be required during the preprocessing stage.

In [21]:
df.duplicated().sum()

np.int64(0)

## Duplicate Rows

No duplicate records were found in the dataset. Each row represents a unique employee record.

In [22]:
df.nunique()

,0
Age,43
Attrition,2
BusinessTravel,3
DailyRate,886
Department,3
DistanceFromHome,29
Education,5
EducationField,6
EmployeeCount,1
EmployeeNumber,1470


In [23]:
print("Missing Values:", df.isnull().sum().sum())
print("Duplicate Rows:", df.duplicated().sum())

constant_columns = df.columns[df.nunique() == 1]
print("Constant Columns:", list(constant_columns))

Missing Values: 0
Duplicate Rows: 0
Constant Columns: ['EmployeeCount', 'Over18', 'StandardHours']


## Constant Columns

The following columns contain only a single unique value:

- EmployeeCount
- Over18
- StandardHours

Since these features provide no variability, they do not contribute to predicting employee attrition and will likely be removed during preprocessing.


## Columns with Very Low Variance

Some features have relatively few unique values, such as:

- PerformanceRating
- StockOptionLevel
- JobLevel
- WorkLifeBalance

These features exhibit limited variability; however, they represent meaningful ordinal information and will be evaluated further during feature analysis rather than being removed solely based on their number of unique values.

## Potential Data Quality Issues

Based on the initial inspection, the following observations were made:

- No missing values are present.
- No duplicate records were found.
- Three features (EmployeeCount, Over18, and StandardHours) contain constant values and are unlikely to contribute to model performance.
- EmployeeNumber is a unique identifier and should not be used as a predictive feature.
- Several categorical and ordinal variables will require appropriate encoding during preprocessing.
- No obvious inconsistencies or invalid data entries were identified during the initial inspection.